phpBB/phpbb/filesystem/filesystem.php

```diff
@@ -0,0 +1,876 @@
+		if ((!$this->is_absolute_path($path) && $realpath === $path) || $realpath === false)
+		{
+			return $this->phpbb_own_realpath($path);
+		}
+
+		// Check for DIRECTORY_SEPARATOR at the end (and remove it!)
+		if (substr($realpath, -1) === DIRECTORY_SEPARATOR)
+		{
+			$realpath = substr($realpath, 0, -1);
+		}
+
+		return $realpath;
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function remove($files)
+	{
+		try
+		{
+			$this->symfony_filesystem->remove($files);
+		}
+		catch (\Symfony\Component\Filesystem\Exception\IOException $e)
+		{
+			// Try to recover filename
+			// By the time this is written that is at the end of the message
+			$error = trim($e->getMessage());
+			$file = substr($error, strrpos($error, ' '));
+
+			throw new filesystem_exception('CANNOT_DELETE_FILES', $file, $e);
+		}
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function rename($origin, $target, $overwrite = false)
+	{
+		try
+		{
+			$this->symfony_filesystem->rename($origin, $target, $overwrite);
+		}
+		catch (\Symfony\Component\Filesystem\Exception\IOException $e)
+		{
+			$msg = $e->getMessage();
+			$filename = substr($msg, strpos($msg, '"'), strrpos($msg, '"'));
+
+			throw new filesystem_exception('CANNOT_RENAME_FILE', $filename, $e);
+		}
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function symlink($origin_dir, $target_dir, $copy_on_windows = false)
+	{
+		try
+		{
+			$this->symfony_filesystem->symlink($origin_dir, $target_dir, $copy_on_windows);
+		}
+		catch (\Symfony\Component\Filesystem\Exception\IOException $e)
+		{
+			throw new filesystem_exception('CANNOT_CREATE_SYMLINK', $origin_dir, $e);
+		}
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function touch($files, $time = null, $access_time = null)
+	{
+		try
+		{
+			$this->symfony_filesystem->touch($files, $time, $access_time);
+		}
+		catch (\Symfony\Component\Filesystem\Exception\IOException $e)
+		{
+			// Try to recover filename
+			// By the time this is written that is at the end of the message
+			$error = trim($e->getMessage());
+			$file = substr($error, strrpos($error, ' '));
+
+			throw new filesystem_exception('CANNOT_TOUCH_FILES', $file, $e);
+		}
+	}
+
+	/**
+	 * phpBB's implementation of is_writable
+	 *
+	 * @todo Investigate if is_writable is still buggy
+	 *
+	 * @param string	$file	file/directory to check if writable
+	 *
+	 * @return bool	true if the given path is writable
+	 */
+	protected function phpbb_is_writable($file)
+	{
+		if (file_exists($file))
+		{
+			// Canonicalise path to absolute path
+			$file = $this->realpath($file);
+
+			if (is_dir($file))
+			{
+				// Test directory by creating a file inside the directory
+				$result = @tempnam($file, 'i_w');
+
+				if (is_string($result) && file_exists($result))
+				{
+					unlink($result);
+
+					// Ensure the file is actually in the directory (returned realpathed)
+					return (strpos($result, $file) === 0) ? true : false;
+				}
+			}
+			else
+			{
+				$handle = @fopen($file, 'r+');
```

Comment:
could be better to use 'w' because a file can be writable but not readable.

Someone knows why it was done like that?